lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java

```diff
@@ -279,7 +297,130 @@ public void collect(int doc) throws IOException {
+
+      final LeafFieldComparator[] comparators = queue.getComparators(context);
+      final int[] reverseMul = queue.getReverseMul();
+      final Sort indexSort = context.reader().getMetaData().getSort();
+      final boolean canEarlyTerminate = canEarlyTerminate(sort, indexSort);
+
+      return new EarlyTerminatingMultiComparatorLeafCollector(comparators, reverseMul, this) {
+
+        boolean collectedAllCompetitiveHits = false;
+
+        @Override
+        public void setScorer(Scorable scorer) throws IOException {
+          super.setScorer(scorer);
+          updateMinCompetitiveScore(scorer);
+        }
+
+        @Override
+        public void collect(int doc) throws IOException {
+
+          if (globalNumberOfHits.incrementAndGet() > numHits) {
+            if (addedSelfToGlobalQueue == false) {
+              Entry returnedEntry = earlyTerminatingFieldCollectorManager.addCollectorToGlobalQueue(earlyTerminatingFieldCollector, docBase);
+
+              if (returnedEntry != null) {
+                filterCompetitiveHit(returnedEntry.doc, false, returnedEntry.values);
+
+                if (queue.size() > 0) {
+                  Entry entry = queue.pop();
+
+                  while (entry != null) {
+                    filterCompetitiveHit(entry.doc, false, entry.values);
+                    entry = queue.pop();
+                  }
+                }
+              }
+              addedSelfToGlobalQueue = true;
+            }
+
+            filterCompetitiveHit(doc, true, comparator.leafValue(doc));
+          } else {
+            // Startup transient: queue hasn't gathered numHits yet
+            int slot = totalHits;
+            ++totalHits;
+
+            comparator.copy(slot, doc);
+            add(slot, doc, comparator.leafValue(doc));
+          }
+        }
+
+        private void filterCompetitiveHit(int doc, boolean doEarlyTermination, Object value) throws IOException {
+          if (collectedAllCompetitiveHits || earlyTerminatingFieldCollectorManager.compareAndUpdateBottom(docBase, doc, value) <= 0) {
+            // since docs are visited in doc Id order, if compare is 0, it means
+            // this document is largest than anything else in the queue, and
+            // therefore not competitive.
+            if (canEarlyTerminate) {
+              if ((globalNumberOfHits.getAcquire() > totalHitsThreshold) && doEarlyTermination) {
+                totalHitsRelation = Relation.GREATER_THAN_OR_EQUAL_TO;
+                throw new CollectionTerminatedException();
+              } else {
+                collectedAllCompetitiveHits = true;
+              }
+            } else if (totalHitsRelation == Relation.EQUAL_TO) {
+              // we just reached totalHitsThreshold, we can start setting the min
+              // competitive score now
+              updateMinCompetitiveScore(scorer);
+            }
+            return;
+          }
+
+          updateMinCompetitiveScore(scorer);
```

Comment:
Why can we `updateMinCompetitiveScore` here?  The global queue may not yet be full at this point?